### ISSUE TITLE
Build System spec should allow lower-level API for package metadata without a file system

### ISSUE THREAD

**Issue by jaraco** [pypa/packaging-problems#226]
In considering #224 and related use-cases, the build system spec defined in [PEP 517](https://www.python.org/dev/peps/pep-0517/) does nominally describe a process for generating metadata for an unbuilt package, but because the process relies on directories and files in a filesystem, it's clumsy for an API to work with this data. It requires the user invoking the process to have access to write to a file system, to allocate a folder to place that data, and later to clean it up.

It should instead be possible for a build backend to be able to provide all of the package's metadata directly, either as a data stream (pipe, for command-line use) or in-memory (for an API).

Such an interface could be provided as a lower-level interface to the `prepare_metadata_for_build_wheel`, where first the metadata is prepared in-memory and then written out to the relevant directory, but the lower-level call would not require a directory in which to write the output. Consider something like `prepare_metadata_for_build_wheel_data`.

This optional hook, if supplied, would be preferred to `prepare_metadata_for_build_wheel` and would  be used to implement `prepare_metadata_for_build_wheel` by writing that content to the file system.

The lower_level `_data` interface, because the metadata is necessarily a tree of files, would need to be a ZipFile in memory or a serialized zip file (or similar).

The advantage of such an interface is that another client of this hook could load and consume the metadata without ever touching the file system... and a build backend could similarly produce this metadata without touching the file system (other than reading the package itself). A client could using something like the [WheelDistribution from importlib_metadata](https://gitlab.com/python-devs/importlib_metadata/blob/05750f6ebac09c4055b319ccabacf098347b87ec/importlib_metadata/_hooks.py#L128) to load and parse the metadata in memory.

As it's currently implemented, both the build backends and the consumers have to be aware of the metadata in the file system and manage the challenges around managing these artifacts across producer/consumer boundaries.

**Comment by jaraco** [comment]
@njsmith @takluyver @ncoghlan I realize what I'm proposing herein is an iteration on the work in PEP 517. To what extent can that standard evolve outside of the PEP process (or conversely is a PEP in order)?

**Comment by pfmoore** [comment]
In principle, packaging interoperability standards like PEP 517 are handled via a streamlined version of the standard PEP process (with myself as BDFL-delegate). We've been moving towards having the master specs as part of the packaging guide rather than as PEPs, but that's an in-progress change at the moment, and for PEP 517 in particular, the PEP is still the reference.

Furthermore, with Python's governance currently being in flux, the situation isn't that clear right now.

My hope is that packaging standards remain independently managed, continuing in the direction we were heading and being managed purely via distutils-sig and documented in the packaging guide<sup>1</sup> - that's the way we were heading, and I don't see any reason why the new Python governance would object.

In practical terms, I'd say that any change or update to PEP 517 of the sort of size you're suggesting<sup>2</sup> should start with a proposal on distutils-sig, where the details can be thrashed out. Once we have a consensus on what changes are needed, we can decide whether it's better done as a modification to PEP 517, or whether it needs a supplementary PEP of its own (or whether we can just move the definitive spec into the packaging guide).

<sup>1</sup> Personally, I think it might actually be better to have our own series of numbered "specs", *referenced* from the guide, but that's mostly just a part-formed thought at the moment.
<sup>2</sup> Minor changes can potentially just be done as PRs against the PEP, but this is too big for that.

**Comment by pfmoore** [comment]
Addressing the details of what you're proposing, a number of thoughts occur to me:

1. Backends have to be run in a separate process (see [the pep](https://www.python.org/dev/peps/pep-0517/#build-environment), which says "Frontends should call each hook in a fresh subprocess"). So transferring data from the backend to the frontend has to work across process boundaries. In general, IPC is tricky to manage in a cross-platform way - a file-based approach is a pretty robust approach, and that was (from what I recall) the key factor in choosing it.
2. If you define the interface in terms of a binary blob of data in zipfile format, consumers still have to wrap that binary blob in a BytesIO stream and open it as a zipfile. There's still complexity for the client - possibly more than there is in a file-based interface (I favour making the interface simple for clients even if it's more complex for backends, not the other way around).
3. If you wanted to, you could write a helper function (which could sit in the `pep517` library) which hid the management of temporary directories from the client. Do you need to change the underlying protocol to get an improved client API? (As noted, I have less interest in simplifying the backend side).

These points should probably be addressed on distutils-sig so that they can be reviewed by a wider audience.

**Comment by jaraco** [comment]
tl;dr: I'm withdrawing this request.

> These points should probably be addressed on distutils-sig so that they can be reviewed by a wider audience.

I struggle with distutils-sig. I already get 100s of e-mails a day, so to subscribe to a mailing list to have a single conversation is not really viable for me. I'm actually subscribed to the list, but it's routed to a special folder that I rarely get to. I wish there were a forum like GitHub issues or similar where I can subscribe to a single conversation.

> Backends have to be run in a separate process.

That clarification does adjust my expectations. It invalidates my expectation that a process could readily call the hook and receive the results in-process. That combined with the fact that subsequent phases of a build (producing build artifacts including metadata) would still require a file system, it seems somewhat clumsy to first produce that data in-memory before extracting it to disk. Although I wish it were possible to retrieve the metadata without requiring a file system, it's probably not worth the trouble.

> consumers still have to wrap that binary blob in a BytesIO stream and open it as a zipfile.

Which is less complex from a state-management point of view, as everything is managed by memory management and garbage collection. There's reduced concern about disk space, access controls, locking semantics, naming concerns, and race conditions.

I was hoping that by presenting the content at a lower level, it would provide an option for services that wished to avoid the issues around file systems.

> Do you need to change the underlying protocol to get an improved client API?

Not necessarily. It's probably best to start with that and see if the concerns I imagine are realized.

Given all of this, I'll withdraw this request. Thanks for the thorough and friendly response.

**Comment by dstufft** [comment]
> I struggle with distutils-sig. I already get 100s of e-mails a day, so to subscribe to a mailing list to have a single conversation is not really viable for me. I'm actually subscribed to the list, but it's routed to a special folder that I rarely get to. I wish there were a forum like GitHub issues or similar where I can subscribe to a single conversation.

There is the packaging category of discuss.python.org! It doesn't have a lot of traffic though and it's pretty new, and not everyone is there so some things still would require talking on distutils-sig (this one probably would be, since it would fall under Paul's "jurisdiction" and he's expressed a desire to have that stuff still in distutils-sig).

But in general, the packaging category on discuss.python.org is there and usable and has this functionality, and more people engaged and discussing things there would be great!

**Comment by pfmoore** [comment]
>  he's expressed a desire to have that stuff still in distutils-sig

As I've used Discourse more, I'm warming to it somewhat. It's much less immediate than a mailing list (I can go days without reading or responding to posts) but that may be as much a positive as a negative factor 😉 

What I do have a strong opinion about, is that wherever discussions on packaging standards take place is *discoverable*. Whatever its flaws, people know, or can find out, about distutils-sig. Discourse, github issues, or any other discussion areas, simply aren't (yet) places that interested parties can easily find. Too often I hear comments from people who think that packaging decisions get made without sufficient community feedback or involvement - that's typically not true, but it's an impression we need to actively fight against. But that's about standards - general discussion can (and should) happen wherever participants are most comfortable.

**Comment by ncoghlan** [comment]
(Following the "Where to discuss things..." meta-tangent)

FWIW, I think there are enough folks active in newer parts of the packaging ecosystem (e.g. PyPUG, pipenv, warehouse) that *aren't* active on distutils-sig that even discussions held there aren't necessarily going to reach everyone we want to reach.

For the CPython governance discussions, the authors of the various PEPs wanted to use discuss.python.org to collect review feedback, so the compromise used to reach those folks that were only following python-committers was to ping the mailing list with links to the web discussions, as well as summaries to help folks decide whether they needed to click through.

Similar to @jaraco, I'm also to the point where I trawl my open source mailing list folders far less frequently than I used to, so the "single interface for all communications" aspect isn't as important for me as it was when I worked for RH.

**Comment by pfmoore** [comment]
OK, those are good points. Let's put it this way. If someone wants to propose an alternative discussion forum for standards (either as a global policy or just on a case by case basis) I have no objection - I withdraw my preference for keeping things on distutils-sig.

Things I *do* think are important:

1. There be a clearly defined, single place for the core discussion. I don't think it's reasonable to ask people to check multiple places for the full picture. By all means discuss anywhere you like, but be prepared to summarise anything said elsewhere back to the main forum (or be prepared for it to be ignored).
2. There needs to be a plan for publicising where the discussion is taking place (like with the governance PEPs, posting reminders to the list), so that people don't feel that they were left out.

I'm happy to leave it at that for now - let's decide the details when we have an actual proposal that needs discussion.

**Comment by pfmoore** [comment]
Possibly, but I don't think it's *too* much of a hardship to expect someone starting a standards discussion to put out a properly worded notification.

The problem with auto-notifying is that we get *all* the topics publicised, which pushes us back to the problem where people miss that something important is happening because they've become blind to the notifications.
